### ISSUE TITLE
Errors_impl-NotFoundError-Undefined symbol-fused_conv2d_bias_activation_op

### ISSUE THREAD

**Issue by mapicccy** [tensorflow/tensorflow#15621]
**What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?**
None.

### System information
- **Operating System**: CentOS 7.3
- **Linux Kernel**: 3.10.0-514
- **TensorFlow installed from source**: https://github.com/tensorflow/tensorflow/
- **TensorFlow version**: 1.4.0
- **Python version**: Python 2.7.5
- **The output of bazel version**:
```
Build target: bazel-out/k8-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Thu Jan 01 00:00:00 1970 (0)
Build timestamp: Thu Jan 01 00:00:00 1970 (0)
Build timestamp as int: 0
```
- **GCC/Compiler version (if compiling from source)**: gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-11)
- **CUDA/cuDNN version**: cuda 8.0/cudnn 7.0
- **GPU model and memory**: NVIDA Tesla P4

### Describe the problem
When exec fused_conv2d_bias_activation_op.py or import the package using `import tensorflow.contrib.fused_conv.python.ops` ，'undefined symbol' error occurs

The source code file path is https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/fused_conv/python/ops/fused_conv2d_bias_activation_op.py

The output of `python fused_conv2d_bias_activation_op_test.py` is as follows:
`tensorflow.python.framework.errors_impl.NotFoundError: /usr/lib/python2.7/site-packages/tensorflow/contrib/fused_conv/python/ops/_fused_conv2d_bias_activation_op.so: undefined symbol: _ZN10tensorflow7functor8PadInputIN5Eigen9GpuDeviceEiiLi4EEclERKS3_NS2_9TensorMapINS2_6TensorIKiLi4ELi1EiEELi16ENS2_11MakePointerEEERKSt5arrayIiLm2EESG_NS7_INS8_IiLi4ELi1EiEELi16ESB_EENS_12TensorFormatE`

It seems that something wrong when loading Eigen library

### The other logs
The output of `objdump -t /usr/lib/python2.7/site-packages/tensorflow/contrib/fused_conv/python/ops/_fused_conv2d_bias_activation_op.so | grep UND | grep Eigen` is as follows:
`0000000000000000       F *UND*  0000000000000000              _ZN10tensorflow25BrainPadding2EigenPaddingENS_7PaddingE
0000000000000000         *UND*  0000000000000000              _ZN10tensorflow7functor8PadInputIN5Eigen9GpuDeviceEiiLi4EEclERKS3_NS2_9TensorMapINS2_6TensorIKiLi4ELi1EiEELi16ENS2_11MakePointerEEERKSt5arrayIiLm2EESG_NS7_INS8_IiLi4ELi1EiEELi16ESB_EENS_12TensorFormatE
0000000000000000         *UND*  0000000000000000              _ZN10tensorflow7functor8PadInputIN5Eigen9GpuDeviceEfiLi4EEclERKS3_NS2_9TensorMapINS2_6TensorIKfLi4ELi1EiEELi16ENS2_11MakePointerEEERKSt5arrayIiLm2EESG_NS7_INS8_IfLi4ELi1EiEELi16ESB_EENS_12TensorFormatE
0000000000000000       F *UND*  0000000000000000              _ZNK10tensorflow15OpKernelContext12eigen_deviceIN5Eigen9GpuDeviceEEERKT_v
0000000000000000         *UND*  0000000000000000              _ZN5Eigen8internal14TensorExecutorIKNS_14TensorAssignOpINS_9TensorMapINS_6TensorIfLi4ELi1EiEELi16ENS_11MakePointerEEEKNS_17TensorReshapingOpIKNS_6DSizesIiLi4EEEKNS_17TensorShufflingOpIKNS9_IiLi3EEEKNS8_ISE_KNS3_INS4_IKfLi4ELi1EiEELi16ES6_EEEEEEEEEENS_9GpuDeviceELb0EE3runERSQ_RKSR_
0000000000000000         *UND*  0000000000000000              _ZN10tensorflow7functor10NHWCToNCHWIN5Eigen9GpuDeviceEfLi4EEclERKS3_NS2_9TensorMapINS2_6TensorIKfLi4ELi1ElEELi16ENS2_11MakePointerEEENS7_INS8_IfLi4ELi1ElEELi16ESB_EE`

**Comment by TianXIA88** [comment]
I have tried Tensorflow 1.3.0 the same "fused_conv2d_bias_activation_op" from

**Comment by TianXIA88** [comment]
I have tried Tensorflow 1.3.0 (built from source), and the same "fused_conv2d_bias_activation_op" from
`import tensorflow as tf
import tensorflow.contrib.fused_conv`

and the same issue occurs with error:
`Traceback (most recent call last):
  File "/mnt/sdb/xt/machine_learning/project/fused_conv_test.py", line 2, in <module>
    import tensorflow.contrib.fused_conv
  File "/usr/lib/python2.7/site-packages/tensorflow/contrib/fused_conv/__init__.py", line 22, in <module>
    from tensorflow.contrib.fused_conv.python.ops.fused_conv2d_bias_activation_op import *
  File "/usr/lib/python2.7/site-packages/tensorflow/contrib/fused_conv/python/ops/fused_conv2d_bias_activation_op.py", line 26, in <module>
    resource_loader.get_path_to_datafile("_fused_conv2d_bias_activation_op.so"))
  File "/usr/lib/python2.7/site-packages/tensorflow/contrib/util/loader.py", line 55, in load_op_library
    ret = load_library.load_op_library(path)
  File "/usr/lib/python2.7/site-packages/tensorflow/python/framework/load_library.py", line 56, in load_op_library
    lib_handle = py_tf.TF_LoadLibrary(library_filename, status)
  File "/usr/lib64/python2.7/contextlib.py", line 24, in __exit__
    self.gen.next()
  File "/usr/lib/python2.7/site-packages/tensorflow/python/framework/errors_impl.py", line 466, in raise_exception_on_not_ok_status
    c_api.TF_GetCode(status))
tensorflow.python.framework.errors_impl.NotFoundError: /usr/lib/python2.7/site-packages/tensorflow/contrib/fused_conv/python/ops/_fused_conv2d_bias_activation_op.so: undefined symbol: _ZN5Eigen8internal14TensorExecutorIKNS_14TensorAssignOpINS_9TensorMapINS_6TensorIfLi4ELi1EiEELi16ENS_11MakePointerEEEKNS_17TensorReshapingOpIKNS_6DSizesIiLi4EEEKNS_17TensorShufflingOpIKNS9_IiLi3EEEKNS8_ISE_KNS3_INS4_IKfLi4ELi1EiEELi16ES6_EEEEEEEEEENS_9GpuDeviceELb0EE3runERSQ_RKSR_`

the missing symbol is demangled as:
Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 4, 1, int>, 16, Eigen::MakePointer>, Eigen::TensorReshapingOp<Eigen::DSizes<int, 4> const, Eigen::TensorShufflingOp<Eigen::DSizes<int, 3> const, Eigen::TensorReshapingOp<Eigen::DSizes<int, 3> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, int>, 16, Eigen::MakePointer> const> const> const> const> const, Eigen::GpuDevice, false>::run(Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 4, 1, int>, 16, Eigen::MakePointer>, Eigen::TensorReshapingOp<Eigen::DSizes<int, 4> const, Eigen::TensorShufflingOp<Eigen::DSizes<int, 3> const, Eigen::TensorReshapingOp<Eigen::DSizes<int, 3> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, int>, 16, Eigen::MakePointer> const> const> const> const> const&, Eigen::GpuDevice const&)

Does this mean that the **_fused_conv2d_bias_activation_op.so** is incorrectly built or lacks some deps in its BUILD file?

**Comment by bignamehyp** [comment]
@yzhwang can you please take a look? Thanks

**Comment by TianXIA88** [comment]
I have worked out a workaround for this issue, and I will post it here in case someone wants to use this fused_conv2d_bias_activation_op:
First add addtional link opt to the BUILD file in tensorflow/contrib/fused_conv/BUILD:
`tf_custom_op_library(
    name = "python/ops/_fused_conv2d_bias_activation_op.so",
    srcs = [
        "kernels/fused_conv2d_bias_activation_op.cc",
        "kernels/fused_conv2d_bias_activation_op.h",
        "kernels/fused_conv_ops_gpu.h",
        "ops/fused_conv2d_bias_activation_op.cc",
    ],
    deps = [
        "//tensorflow/core:lib_proto_parsing",
        "//tensorflow/core/kernels:bounds_check_lib",
        "//tensorflow/core/kernels:conv_2d_hdrs",
        "//tensorflow/core/kernels:conv_ops_gpu_hdrs",
        "//tensorflow/core/kernels:gpu_util_hdrs",
        "//tensorflow/core/kernels:ops_util_hdrs",
    ],
    linkopts = ["-shared-libgcc /usr/lib/python2.7/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so -Xlinker --unresolved-symbols=ignore-in-shared-libs"],
)`
This is to explicitly link against the _pywrap_tensorflow_internal.so since some symbols are hidden by Bazel in new version (from [Issue 15582](https://github.com/tensorflow/tensorflow/issues/15582). And this "--unresolved-symbols=ignore-in-shared-libs" option will help you get the line that causes the "undefined symbol" error.

After this is done, use bazel to rebuild the op and you can find out the reason causing the problem:
`bazel-out/local_linux-opt/bin/tensorflow/contrib/fused_conv/_objs/python/ops/_fused_conv2d_bias_activation_op.so/tensorflow/contrib/fused_conv/kernels/fused_conv2d_bias_activation_op.pic.o: In function `tensorflow::LaunchFusedConv2DBiasActivationOp<Eigen::GpuDevice, float, float, float>::launch(tensorflow::OpKernelContext*, bool, tensorflow::Tensor const&, float, tensorflow::Tensor const&, int, int, Eigen::PaddingType const&, tensorflow::Tensor const&, float, tensorflow::Tensor const&, tensorflow::ActivationMode, tensorflow::TensorFormat, tensorflow::FilterTensorFormat, tensorflow::Tensor*)':
fused_conv2d_bias_activation_op.cc:(.text._ZN10tensorflow33LaunchFusedConv2DBiasActivationOpIN5Eigen9GpuDeviceEfffE6launchEPNS_15OpKernelContextEbRKNS_6TensorEfS8_iiRKNS1_11PaddingTypeES8_fS8_NS_14ActivationModeENS_12TensorFormatENS_18FilterTensorFormatEPS6_[_ZN10tensorflow33LaunchFusedConv2DBiasActivationOpIN5Eigen9GpuDeviceEfffE6launchEPNS_15OpKernelContextEbRKNS_6TensorEfS8_iiRKNS1_11PaddingTypeES8_fS8_NS_14ActivationModeENS_12TensorFormatENS_18FilterTensorFormatEPS6_]+0x1b5b): undefined reference to `Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 4, 1, int>, 16, Eigen::MakePointer>, Eigen::TensorReshapingOp<Eigen::DSizes<int, 4> const, Eigen::TensorShufflingOp<Eigen::DSizes<int, 3> const, Eigen::TensorReshapingOp<Eigen::DSizes<int, 3> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, int>, 16, Eigen::MakePointer> const> const> const> const> const, Eigen::GpuDevice, false>::run(Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 4, 1, int>, 16, Eigen::MakePointer>, Eigen::TensorReshapingOp<Eigen::DSizes<int, 4> const, Eigen::TensorShufflingOp<Eigen::DSizes<int, 3> const, Eigen::TensorReshapingOp<Eigen::DSizes<int, 3> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, int>, 16, Eigen::MakePointer> const> const> const> const> const&, Eigen::GpuDevice const&)'
`
Then we go to the "tensorflow/contrib/fused_conv/kernels/fused_conv2d_bias_activation_op.cc" file and **comment out** the GPU-Float specific:
`// Registration of the GPU implementations.
REGISTER_KERNEL_BUILDER(
    Name("FusedConv2DBiasActivation")
        .Device(DEVICE_GPU)
        .TypeConstraint<float>("T")
        .TypeConstraint<float>("Tbias")
        .HostMemory("conv_input_scale")
        .HostMemory("side_input_scale"),
    FusedConv2DBiasActivationOp<GPUDevice, float, float, float>);
`

Then you can rebuild the project and then the FusedConv2DBiasActivation should be OK to used, But only the INT8 type input is allowed.

To sum up, it seems the float type  GPU declaration some how hit a library-missing issue. However I haven't figured it out...

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activityand the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by TianXIA88** [comment]
Is there any news about this issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 128 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 143 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 158 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 173 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 188 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 203 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by aselle** [comment]
I apologize, but nobody is working on this, and since it is in contrib, it is not officially supported.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by tensorflowbutler** [comment]
As this issue has invited community support, please remove the assignee. Otherwise, remove the `community support` label. Thank you.

**Comment by ymodak** [comment]
Closing this issue due to staleness. Feel free to reopen when new information is available. Thanks!
